src/parser.rs

```diff
@@ -114,14 +114,21 @@ fn parse_err(args: &[u8]) -> ControlOp {
     ControlOp::Err(err_description.to_string())
 }
 
-pub(crate) fn expect_info(reader: &mut BufReader<TcpStream>) -> io::Result<ServerInfo> {
-    let op = parse_control_op(reader)?;
+pub(crate) fn expect_info(reader: &mut TcpStream) -> io::Result<ServerInfo> {
+    let mut buf = Vec::with_capacity(512);
 
-    if let ControlOp::Info(info) = op {
-        Ok(info)
-    } else {
-        Err(Error::new(ErrorKind::Other, "INFO proto not found"))
+    while !buf.ends_with(b"\r\n") {
+        let byte = &mut [0];
+        reader.read_exact(byte)?;
+        buf.push(byte[0]);
+    }
+
+    if buf.starts_with(b"INFO ") {
+        if let Ok(info) = serde_json::from_slice(&buf[b"INFO ".len()..]) {
+            return Ok(info);
+        }
     }
+    Err(Error::new(ErrorKind::Other, "INFO proto not found"))
```

Comment:
This just loops over the initial INFO block one byte at a time. There are faster ways of doing this, but I assume this happens so infrequently that performance isn't really an issue here. Subsequent INFO blocks that get passed over the possibly-TLS-upgraded connection are parsed in a more efficient way.

Reply:
I think that is fine.